File: 剑指offer/36. 二叉搜索树与双向链表.py
"""
执行用时 :
32 ms
, 在所有 Python3 提交中击败了
98.58%
的用户
内存消耗 :
14.9 MB
, 在所有 Python3 提交中击败了
100.00%
的用户
"""

class Solution:
    def treeToDoublyList(self, root: 'Node') -> 'Node':
        def dfs(cur):
            if not cur: return
            dfs(cur.left) # 递归左子树
            self.pre.right, cur.left = cur, self.pre # 修改引用
            self.pre = cur # 保存 cur
            dfs(cur.right) # 递归右子树
        
        if not root: return
        head = self.pre = ListNode(0) # 伪头节点
        dfs(root)
        head = head.right
        head.left, self.pre.right = self.pre, head
        return head


https://leetcode-cn.com/problems/er-cha-sou-suo-shu-yu-shuang-xiang-lian-biao-lcof/solution/mian-shi-ti-36-er-cha-sou-suo-shu-yu-shuang-xian-5/

File: 剑指offer/C++_Version/36. 二叉搜索树与双向链表.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;

    Node() {}

    Node(int _val) {
        val = _val;
        left = NULL;
        right = NULL;
    }

    Node(int _val, Node* _left, Node* _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};
*/
class Solution {
    
public:
    Node* treeToDoublyList(Node* root) 
    {
        if(!root) return NULL;
        dfs(root);
        pre->right = head;
        head->left = pre;
        return head;
    }

private:
    Node *pre, *head;
    void dfs(Node* cur)
    {
        if(!cur) return;
        dfs(cur->left);
        if(pre)
        {
            pre->right = cur;
            cur->left = pre;
        }
        else
        {
            head = cur;
        }
        pre = cur;
        dfs(cur->right);
    }

};
